src/rust/derive-dynamic-node/src/lib.rs

```diff
@@ -206,7 +214,51 @@ pub fn derive_grapl_session(input: TokenStream) -> TokenStream {
     q.into()
 }
 
-fn property_methods(property_name: &Ident, property_type: &Type) -> TS2 {
+fn identity_prop_setter(field: &Field, property_name: &Ident) -> TS2 {
+    let mut created_time_prop = false;
+    let mut last_seen_time_prop = false;
+    let mut terminated_time_prop = false;
+
+    for attr in field.attrs.iter() {
+        on_grapl_attrs(attr, |meta_attr| {
+            created_time_prop |= meta_attr == CREATE_TIME;
+            last_seen_time_prop |= meta_attr == LAST_SEEN_TIME;
+            terminated_time_prop |= meta_attr == TERMINATE_TIME;
+        });
+        if [
+            &created_time_prop,
+            &last_seen_time_prop,
+            &terminated_time_prop,
+        ]
+        .iter()
+        .any(|b| **b)
+        {
+            break;
+        }
+    }
+
+    let ident = match (created_time_prop, last_seen_time_prop, terminated_time_prop) {
+        (true, _, _) => syn::Ident::new(&CREATE_TIME, field.span()),
+        (_, true, _) => syn::Ident::new(&LAST_SEEN_TIME, field.span()),
+        (_, _, true) => syn::Ident::new(&TERMINATE_TIME, field.span()),
+        _ => return quote!(),
+    };
+    quote!(
+        let mut self_strategy = mut_self.id_strategy[0].strategy.as_mut().unwrap();
+        match self_strategy {
+            grapl_graph_descriptions::graph_description::id_strategy::Strategy::Session(
+                grapl_graph_descriptions::graph_description::Session{ref mut #ident, ..}
+            ) => {
+                * #ident = #property_name;
+            }
+            s => panic!("Can not set timestamps on non-Session strategy {:?}", s)
```

Comment:
I think ultimately there's nothing a caller can do in this case except log and panic anyway - this indicates a bug - as you said, most likely some sort of version mismatch. I don't think retrying really would come up here, since the only way to fix the code would be to redeploy a new version.

Reply:
Do we have infrastructural guarantees that we would not lose data here, if it panics?

Reply:
Yeah, it'll get retried.